core/packages/contracts/src/TokenVault.sol

```diff
@@ -0,0 +1,55 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity ^0.8.19;
+
+import "openzeppelin/access/AccessControl.sol";
+import "openzeppelin/token/ERC20/IERC20.sol";
+import "openzeppelin/token/ERC20/utils/SafeERC20.sol";
+
+/// @title ERC20 Vault
+/// @dev Holds ERC20 Tokens on behalf of ERC20App.
+contract TokenVault is AccessControl {
+    using SafeERC20 for IERC20;
+
+    /// @dev Emitted when funds are deposited.
+    event Deposit(address sender, address token, uint256 amount);
+
+    /// @dev Emitted when funds are withdrawn.
+    event Withdraw(address recipient, address token, uint256 amount);
+
+    /// @dev Not enough funds to transfer.
+    error InsufficientBalance();
+
+    bytes32 public constant ADMIN_ROLE = keccak256("ADMIN_ROLE");
+    bytes32 public constant WITHDRAW_ROLE = keccak256("WITHDRAW_ROLE");
+    bytes32 public constant DEPOSIT_ROLE = keccak256("DEPOSIT_ROLE");
+
+    /// @dev stores the total balance of each token locked in the vault.
+    mapping(address => uint128) public balance;
+
+    constructor() {
+        _grantRole(ADMIN_ROLE, msg.sender);
```

Comment:
ADMIN_ROLE is set to deployer account, and they can change sender/withdrawal role even after deployment? Should have a revokeRole there in the contract / deployment process?

Reply:
See my latest PR to support upgrades:

https://github.com/Snowfork/snowbridge/pull/801

The deployer's admin privileges are revoked.